### PR TITLE
Add PullRequests stream and get all issues and prs regardless of state

### PR DESCRIPTION
**Pull request by ericboucher**
Following up on #17, I would like to propose this solution.

This PR:
- creates a new PullRequests stream that fetches PRs only (regardless of state).
- updates the Issue stream to fetch all issues (and PRs) regardless of state and add a "type" field

### CODE REVIEW

**Comment by aaronsteers** [comment]
@ericboucher - First, thanks very much for this contribution.

A quick first-round review did not surface any red flags. I'm very happy for issues to include "all" states and to be adding the Pull requests as a new stream.

However, I'm not sure on whether it is best to have the issues stream as a superset - including "normal" issues and also pull requests - versus having it be mutually exclusive with pull requests.

Things I considered:

1. field availability: I'm not sure if the issue fields are a strict subset of those available for pulls - or if there are some value-add fields a user may want that are only available in the issues stream. 
2. ontological argument: from github's perspective (at least in the API), PRs are also Issues. 
3. Singer precedent: I did a quick scan of Stitch's implementation, and it looks like they are including both types in their implementation. Here's [the code I reviewed](https://github.com/singer-io/tap-github/blob/master/tap_github/__init__.py#L908), in case I'm reading it wrong.

What do you think? None of these are showstoppers in themselves and I'm curious of your thoughts here. I'm happy to discuss further here, or we can move that topic to another issue in order to speed along the addition of the PR stream and non-open issue states.

**Comment by ericboucher** [comment]
@aaronsteers

I am not sure either. IMHO the fact that PRs are Issues from a GitHub API standpoint is a historical artifact. So I would be in favor of splitting them more cleanly.

However, from a "tap" standpoint, the goal is to get all the data as is, without too much interpretation. So from a "tap" perspective, it means getting all issues returned by the API, PRs included.

TLDR - I'll defer to your team as the main maintainers and make the changes needed either way :)

**Comment by aaronsteers** [comment]
> @aaronsteers
> 
> I am not sure either. IMHO the fact that PRs are Issues from a GitHub API standpoint is a historical artifact. So I would be in favor of splitting them more cleanly.
> 
> However, from a "tap" standpoint, the goal is to get all the data as is, without too much interpretation. So from a "tap" perspective, it means getting all issues returned by the API, PRs included.
> 
> TLDR - I'll defer to your team as the main maintainers and make the changes needed either way :)

I can see going either way. @ericboucher - Have you checked if the PR stream has all the interesting/relevant fields in the Issues stream?

Would love feedback from other contributors @Ry-DS and @laurentS

**Comment by ericboucher** [comment]
@aaronsteers diving a bit more into it, the PR stream includes more data for PRs as well as the fields included in 'issues'.

Therefore, I think my preference is to have the `Issues` stream pull issues only and not PRs, and the `PullRequest` stream gets PRs with additional info (such as base, target, ...)

Otherwise, it means that we would get the same object (PRs) twice to get all relevant information. Thoughts?

File: tap_github/streams.py
Comment (aaronsteers) [inline tap_github/streams.py:461]: Since PullRequests don't have child stream types any more, I think this method is not used and can be safely removed. (Probably also not used for `IssuesStream` anymore but that's a separate topic.)

**Comment by aaronsteers** [comment]
> @aaronsteers diving a bit more into it, the PR stream includes more data for PRs as well as the fields included in 'issues'.
> 
> Therefore, I think my preference is to have the `Issues` stream pull issues only and not PRs, and the `PullRequest` stream gets PRs with additional info (such as base, target, ...)
> 
> Otherwise, it means that we would get the same object (PRs) twice to get all relevant information. Thoughts?

@ericboucher 
Thanks for comparing those field lists. That's a big help.

The other thing I just realized though is that children - like issue comments - are generically labeled as "Issue Comments" and don't have any data in their record to tell the user whether the comment is actually an Issue comment or a PR comment. The join downstream then would get quite messy to get PR comments linked back up with PRs.

Instead of a simple left join to everything we need, like:

```sql
SELECT
    issues.id,
    issues.type,
    count(*) AS comment_count,
    max(issue_comments.post_date) as latest_comment,
    -- ...
FROM issue_comments
LEFT JOIN issues -- most things can be pull from here
ON ...
LEFT JOIN pull_requests. -- optional join, if PR-related fields are needed
ON ...
GROUP BY 1, 3
```

We'd have to re-union them back together in order to make the join, something like:

```sql
SELECT
    issues.id,
    issues.type,
    count(*) AS comment_count,
    max(issue_comments.post_date) as latest_comment
FROM issue_comments
LEFT JOIN (
    SELECT ... FROM issues 
    UNION ALL 
    SELECT ... FROM pull_requests
) AS all_issue_types
ON ...
GROUP BY 1, 3
```


In addition to the messiness of joins downstream, there are some other weird edge cases which I can imagine causing problems, like a user pulling just the issue_comments stream and issues streams (without selecting pull_requests) and then not having any means of knowing if the parent issue record is missing due to it being a PR or due to another data quality or replication issue.

So, I guess in summary, because children don't know what type of issue their parent is, I still would like to keep the issues list as a superset of all issues even though this is admittedly redundant.

I should add: there's an easy workaround to exclude PRs from the issues stream, which would be to use Stream Maps in the config and set something like:

```js
{
    // ..
    "stream_maps": {
        "issues": {
            "__filter__": "record['type'] = 'issue'"
        }
    }
{
```

Would you be okay keep as-is given the above?

**Comment by ericboucher** [comment]
@aaronsteers works for me. I updated the streams following your input. Should I add the stream_map example in a comment somewhere maybe?

**Comment by aaronsteers** [comment]
> @aaronsteers works for me. I updated the streams following your input. Should I add the stream_map example in a comment somewhere maybe?

Totally optional, but yes, if you think it could be helpful. Certainly it makes sense to call out Issues being inclusive also of Pull Requests. My little config snippet was written by hand from memory so it might take a little tweaking to get syntax just right. 
 😄

**Comment by ericboucher** [comment]
@aaronsteers note that all the children's streams are also available for pulls. So we could also take the approach of getting their children separately. We will probably be getting commits and reviews.

https://docs.github.com/en/rest/overview/endpoints-available-for-github-apps#pulls

**Comment by aaronsteers** [comment]
> @aaronsteers note that all the `comments` stream is also available for pulls. So we could also take the approach of getting pulls children separately.
> 
> https://docs.github.com/en/rest/overview/endpoints-available-for-github-apps#pulls

It appears so at first but I think not actually. The text if you read the description refers to "review comments" which are comments attached to a review. My understanding is that not all PR comments are "review comments".

This section from the docs points to the Issue Comments API for PR comments:

**Comment by ericboucher** [comment]
Way to make this complicated, thanks GitHub! haha

**Comment by ericboucher** [comment]
@aaronsteers I think we are good like this then. To summarize, these are the changes included in this PR:

- IssueStream pulls all issues and PRs regardless of state and we add a "`type` field making it easier to differentiate and filter
- PullRequestStream pulls all PRs regardless of state with the extra info available with the `/pulls` endpoint (base, target, ...)

File: tap_github/streams.py
Comment (aaronsteers) [inline tap_github/streams.py:240]: Looks like this is failing (missing?) for regular issues. What do you think of?:

```suggestion
        row["type"] = "pull_request" if "pull_request" in row else "issue"
```

**Review by aaronsteers — CHANGES_REQUESTED** [review]

**Review by aaronsteers — APPROVED** [review]

**Comment by aaronsteers** [comment]
@ericboucher - Do you mind resolving the conflicts from the recent merge to `main`? Otherwise, this looks ready to go. ✅

**Comment by ericboucher** [comment]
@aaronsteers ready to go :)

**Pull request merged**
